### DISCUSSION TITLE
stb_truetype: stbtt_GetPackedQuad doesn't seem to stick to baseline

### DISCUSSION THREAD

**Discussion by forksnd** [nothings/stb#1723] [Q&A] [answered]
@nothings 
I am trying out the [new texture baking API](https://github.com/nothings/stb/blob/master/stb_truetype.h#L568-L674) by first baking the texture, and then dumping the texture and the `stbtt_packedchar` array in a file. This file gets loaded in another process that uses `stbtt_GetPackedQuad` to render the characters. When I do this, the characters don't stick to the baseline as can be seen in the image below.

![image](https://github.com/user-attachments/assets/4b5d359d-edf8-4d2f-9460-b70243f11410)

The quad is created and rendered as shown in [this example code](https://github.com/nothings/stb/blob/5c205738c191bcb0abc65c4febfa9bd25ff35234/tests/oversample/main.c#L120-L122). The font I am using is [this one](https://github.com/be5invis/Iosevka/releases/download/v32.1.0/PkgTTF-IosevkaEtoile-32.1.0.zip) (the "Regular" variant).

Any idea what I might be doing wrong?

-----

The basic code flow I am using is:

1. **For baking:**
    ````c
    stbtt_InitFont();
    stbtt_PackBegin();
    stbtt_PackSetOversampling();
    stbtt_PackFontRanges(); // multiple times with different fontdata each time
    stbtt_PackEnd();
    fwrite(packedchar[]); // As a header file full of designated initializers (floats printed as hex using %a)
    fwrite(pixels); // As a header file containing an array of uint8_t (like xxd)
    ````
    Font size in `stbtt_pack_range` is set using the `STBTT_POINT_SIZE` macro.
2. **For rendering:**
    ````c
    #include "pixels.h"
    #include "packedchar.h"
    stbtt_GetPackedQuad();
    render(quad);
    ````

**Comment by nothings** [comment]
Impossible to say without more info. The most likely issues would be in your code: either your packedchar printing code is wrong, or your render() code is wrong.

Obviously the first thing to do is to simplfy and eliminate variables. Run stbtt_ in the same process and make sure that works. Write out the packedchars to a binary file with `fwrite` not `fprintf` and make sure that works. If both of those work, then you've narrowed it down to your `fprintf`. If it fails in the earlier steps, you can diagnose why with fewer things to worry about.

**Reply by forksnd** [reply]
Thanks, just wanted to confirm that I am not missing a step, to maybe adjust the resulting quad with some font metrics or something. The entire code will be too complex to post here, I'll see what I can debug.

**Reply by nothings** [reply]
If you're calling stbtt_PackFontRanges multiple times, each call writes out stbtt_packedchar data, so you also need to make sure that's to separate arrays (or non-overlapping parts of the same array).

**Reply by forksnd** [reply, marked as answer]
Fixed! There were two issues:

1. The floats were getting cast to ints deep in the GPU pipeline. (I forgot that the previous font code dealt with pixel-coordinates)
2. I had set `align_to_integer` in `stbtt_GetPackedQuad` to `true`.

Removing the cast and setting `align_to_integer` to `false` seems to have fixed the issue.

![Fixed screenshot](https://github.com/user-attachments/assets/828d91bb-53a9-4e91-85b1-54379ea73211)
